servicemanager/servicemanager.go

```diff
@@ -342,3 +548,114 @@ func (sm *ServiceManager) prepareServiceFS(imagePath string, gid int) (rootFSDig
 
 	return rootFSDigest, nil
 }
+
+func (sm *ServiceManager) removeDamagedServiceFolders(services []ServiceInfo) error {
+	files, err := ioutil.ReadDir(sm.servicesDir)
+	if err != nil {
+		return aoserrors.Wrap(err)
+	}
+
+filesLoop:
+	for _, file := range files {
+		fullPath := path.Join(sm.servicesDir, file.Name())
+
+		for _, service := range services {
+			if fullPath == service.ImagePath {
+				continue filesLoop
+			}
+		}
+
+		if err = os.RemoveAll(fullPath); err != nil {
+			log.Errorf("Can't remove service directory: %v", err)
+		}
+	}
+
+	return nil
+}
+
+func (sm *ServiceManager) tryAllocateSpace(requiredSize int64) (err error) {
+	sm.Lock()
+	defer sm.Unlock()
+
+	if sm.numTryAllocateSpace == 0 {
+		if sm.availableSize, err = GetAvailableSize(sm.servicesDir); err != nil {
+			return aoserrors.Wrap(err)
+		}
+	}
+
+	for sm.availableSize < requiredSize {
+		if len(sm.pendingRemove) == 0 {
+			return aoserrors.New("not enough memory to delete service")
+		}
+
+		freedServiceSize, _, err := sm.removeService(sm.pendingRemove[0])
+		if err != nil {
+			return aoserrors.Wrap(err)
+		}
+
+		sm.availableSize += freedServiceSize
+
+		sm.pendingRemove = sm.pendingRemove[1:]
+	}
+
+	sm.availableSize -= requiredSize
+	sm.numTryAllocateSpace++
+
+	return nil
+}
+
+func (sm *ServiceManager) extractPackageByURL(
+	extractDir, downloadDir, packageURL string, fileInfo image.FileInfo,
+) (serviceSize int64, err error) {
+	urlVal, err := url.Parse(packageURL)
+	if err != nil {
+		return 0, aoserrors.Wrap(err)
+	}
+
+	var sourceFile string
+
+	if urlVal.Scheme != "file" {
+		if sm.isDownloadSamePartition {
+			if err = sm.tryAllocateSpace(int64(fileInfo.Size)); err != nil {
+				return 0, aoserrors.Wrap(err)
+			}
+		}
+
+		defer sm.deallocateSpace(int64(fileInfo.Size))
+
+		if sourceFile, err = image.Download(context.Background(), downloadDir, packageURL); err != nil {
```

Comment:
Same consideration as for layer manager.

Reply:
move to separate  task